### PR TITLE
DHCP DUID and IAID configurability.

### PR DESCRIPTION
**Pull request by vinaykul**
This change implements the earlier proposed capability  to allow configuration of DHCP DUID and IAID values in systemd-networkd.  (Ref: https://github.com/systemd/systemd/issues/394#issuecomment-185489519 )

I have tested this change as follows: 
1. Deploy a VM that uses dhclient to acquire a DHCP IP addresses (DUID-LLT).
2. Reading the DUID and IAID values from /var/lib/dhcpv6/client.leases file, and plug them into DUIDValue and IAIDValue fields on PhotonOS running networkd with this change.
3. Suspend the dhclient VM.
4. Restart networkd based VM and verify that IPv6 address migrated.

With ClientIdentifier=duid, I was able to migrate the IPv4 address from one DUID-DHCP VM to another VM by plugging in DUID and IAID values from /run/systemd/netif/leases/2 file. Wireshark traces confirm what I expected to see.

### CODE REVIEW

File: man/dhcp-duid.xml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+<refentry id="dhcp-duid">
+  <refentryinfo>
+    <title>dhcp-duid</title>
+    <productname>systemd</productname>
+
+    <authorgroup>
+      <author>
+        <contrib>Developer</contrib>
+        <firstname>Lennart</firstname>
```
Comment (medhefgo) [inline man/dhcp-duid.xml]: Are you sure that's your name and email? :P

**Comment by poettering** [comment]
Hmm, I wonder if this should be a per-system or rather a per-network configuration option,

Or maybe it should be per-system _and_ per-network? If so, it would probably make sense to introduce an /etc/systemd/networkd.conf file, which knows a configuration option for this plus a per-network one, in *.network that may be used to override this.

Anyway, I think @teg should really comment on this, @teg?

**Comment by poettering** [comment]
I commented on a few issues, but I think @teg should weigh in regarding where this should actually be configurable. I personally think a generic /etc/systemd/networkd.conf (that is similar to /etc/systemd/resolved.conf or /etc/systemd/timesyncd.conf) is the right choice, plus possibly a per-network overrding option in *.network...

**Comment by teg** [comment]
Overall I like this. I agree with @poettering's suggestion of making this globally configurable though, with a possible override.

Moreover, I really would prefer us to be a bit more structured in how we parse the data. I.e., I'd make it look like this (both globally and possibly per .network):

[DUID]
Type=raw|link-layer|link-layer-and-time|vendor|uuid
UUID=
MACAddress=
TimeStamp=
EnterpriseNumber=
RawData=

With the logic being that the type determines what other fields are optional/mandatory, as follows:
- `raw` requires RawData to be set (then including the full DUID just like your current logic works)
- `link-layer` uses the current mac address, or optionally the one given in `MACAddress`
- `link-layer-and-time` is the same, but it also requires the time to be set (the man page should point out that you probably want to use the @<seconds since UNIX epoch> syntax)
- `vendor` defaults to our own enterprise number, but allows another to be set, it requires `RawData` to fill in the rest.
- `uuid` requires `UUID` to be set.

What do you think?

**Comment by vinaykul** [comment]
Sorry for the delayed response... a couple of other work priorities came up.

I agree with @poettering that /etc/systemd/networkd.conf is the right container for per-machine network configurations (DUID etc..), and right thing to do for the long run. I'm not very familiar with the overall parser design/architecture, started looking into it... looks doable without a steep learning ramp. So I'll plan to work on it.

I don't think we should have a per-network configurability of DUID... I considered adding this alongside IAID, but after looking closely at the RFCs, it appears DUID is a per-machine identity and not a per-network identity, so I chose not to. Unless there is a use-case we know of, we should keep it simple ... having this per-network might just confuse users.

About the DUID structure, while it does offer more flexibility, the added complexity does raise additional cases to consider/handle - for e.g. if MACAddress is not specified, which MAC address to use in a multi-interface system? 

I can start with adding support for the raw-type (immediate and likely the most-common use case), and leave other types for future extension. 

Concerns/questions?

File: man/dhcp-duid.xml (hunk 1 above)
Comment (vinaykul) [inline man/dhcp-duid.xml]: Sharp eyes! I was hoping to slip that by so Lennart gets the flame mails ;) 

I'll fix it in the next update.

**Comment by teg** [comment]
@vinaykul yeah, that sounds perfect. Just keep in mind that we MIGHT want the per-interface stuff one day, or the more structured config, but there is no need to implement that from the get-go. Just doing the stuff necessary to get your raw DUID sounds good to me.

**Comment by nfnty** [comment]
Per-interface configuration would be very useful for multiple physical interfaces and multiple DHCP servers.

**Comment by vinaykul** [comment]
@nfnty Please review RFC 3315 section 9. Under DUID-LLT, the RFC specifies to use the link-layer address of any one of the interfaces in the DHCP device, and that choice of the interface is arbitrary, and further that the same DUID-LLT should be used to configure all interfaces, regardless of which interface address was used to generate the DUID. This implies that a system should have no more than one DUID. So I think we should not have DUID per .network.

That said, it is quite easy to add .network override for DUID. ( I can take it up in a future commit so I don't reset my mostly complete VM upgrade scenario testing on my current implementation I'm working on that allows DUID configuration in networkd.conf file ). 

What is the scenario for which you need this?

**Pull request closed without merging**